### ISSUE TITLE
Detect Proton saves for non-Steam games added in Steam

### ISSUE THREAD

**Issue by Bboy486** [mtkennerly/ludusavi#130]
### Ludusavi version

v0.13.0

### Operating system

Linux (Steam Deck)

### Installation method

Flatpak

### Description

For non steam games is there a way for your tool to locate the saves and backup? I added the folder but it is only finding the games installed in steam (ignores anything that is a non-steam game).

I set the roots for the SD card and also ssd where the games are along with the steam root (which it finds the installed steam games), and heroic locations. However it doesn't detect any of the non steam games.

**Comment by mtkennerly** [comment]
Thanks for your report! I suspect this may be an issue with Flatpak permissions, but let's try to narrow it down.

* Heroic isn't currently supported (#94). Other than Heroic and Steam, what other sources/stores do you have? Could you give me a specific example of a non-Steam/Heroic game that can't be found and where it's installed?
* Could you please upload [your config file](https://github.com/mtkennerly/ludusavi#configuration)? I'd like to see how your roots are configured.
* For whichever game you answered to the first question (let's say `$GAME`), please run this command and then upload [your logs](https://github.com/mtkennerly/ludusavi#logging):

  ```
  RUST_LOG=ludusavi=trace flatpak run com.github.mtkennerly.ludusavi backup --preview "$GAME"
  ```

**Comment by Bboy486** [comment]
I'm not seeing the config under ~/.config/ludusavi and not sure where that other file is. Is the Deck config in a different place?

**Comment by Bboy486** [comment]
I'm not seeing logs where the wiki said or in the path above

RUST_LOG=ludusavi=trace flatpak run com.github.mtkennerly.ludusavi backup --preview "$GAME"
No info for these games:
  -

**Comment by mtkennerly** [comment]
Ah, it didn't get far enough to generate any logs because it didn't understand the title. `$GAME` was just a placeholder - you should change it to `Deadpool`, like this:

```
RUST_LOG=ludusavi=trace flatpak run com.github.mtkennerly.ludusavi backup --preview "Deadpool"
```

Based on [PCGW](https://www.pcgamingwiki.com/wiki/Deadpool), and since the game is in an other-type root, it should find saves at `/home/deck/Games/Deadpool/TransGame/Config/PC/Cooked`. Could you confirm that that folder exists and is not empty?

**Comment by mtkennerly** [comment]
No, you shouldn't make each individual game into a root. That won't work. Roots are for groups of games, not individual games.

Since you have `/home/deck/Games` as an other-type root, that means Ludusavi will automatically look for any games installed inside of it:

* /home/deck/Games/Deadpool
* /home/deck/Games/Celeste
* /home/deck/Games/Steelrising

...and so on. There's more info about roots [in the docs](https://github.com/mtkennerly/ludusavi#roots).

> I can but to be clear /Games is the actual folder with all my non-steam games on both the internal and SD card.

Can you elaborate on this?

* Do you really mean `/Games` (directly under the file system root), or do you mean `/home/deck/Games`?
* How do you have it containing *both* internal and SD card installations? Are you using symlinks? Is this managed by some other launcher/program?
* Can you give me a concrete example of how you installed Deadpool? Please be very specific and detailed. Is it on the internal storage or on the SD card? If `/home/deck/Games/Deadpool` is a symlink, where does it point to? Did you use a launcher like Heroic to install it, and if so, which launcher? All of these questions could affect the Flatpak permissions for Ludusavi to be able to scan that folder.

**Comment by Bboy486** [comment]
I mean home/deck/Games for ssd
and /SD card/Games for SD card or external HDD

No symlinks at all 
So your Deadpool example is correct: home/deck/Games/Deadpool within /Games/ are all of my other non steam games. They are setup by game name like you had in your example. 

Deadpool is on the ssd. I lives in the /deck/Games folder. I add them as non steam games. 

Heroic is also in the /Games/ folders but I'm not focusing on that based on the open issue you linked to previously.

**Comment by mtkennerly** [comment]
Thank you for clarifying!

I do still need you to run this command and provide the logs, please:

```
RUST_LOG=ludusavi=trace flatpak run com.github.mtkennerly.ludusavi backup --preview "Deadpool"
```

**Comment by Bboy486** [comment]
RUST_LOG=ludusavi=trace flatpak run com.github.mtkennerly.ludusav
error: app/com.github.mtkennerly.ludusav/x86_64/master not installed
(

**Comment by mtkennerly** [comment]
That's not the command from my previous comment 😅 Please copy/paste it and run it exactly as-is.

Afterwards, I'll need the log file from the config directory, **not** the command line output.

**Comment by mtkennerly** [comment]
That's not the command from my previous comment 😅 Please copy/paste it and run it exactly as-is.

Afterwards, I'll need the log file from the config folder, **not** the command line output.

**Comment by Bboy486** [comment]
Sorry 😐

RUST_LOG=ludusavi=trace flatpak run com.github.mtkennerly.ludusavi backup --preview "Deadpool"
Overall:
  Games: 0
  Size: 0 B
  Location: /run/media/mmcblk0p1/ludusavi

**Comment by mtkennerly** [comment]
Hmm... Could you try this instead, with Flatpak's own `--env` option? Maybe it just wasn't seeing the environment variable.

```
flatpak run --env=RUST_LOG=ludusavi=trace com.github.mtkennerly.ludusavi backup --preview "Deadpool"
```

**Comment by Bboy486** [comment]
flatpak run --env=RUST_LOG=ludusavi=trace com.github.mtkennerly.ludusavi backup --preview "Deadpool"
Overall:
  Games: 0
  Size: 0 B
  Location: /run/media/mmcblk0p1/ludusavi

**Comment by Bboy486** [comment]
Any ideas about where the logs are located?

**Comment by mtkennerly** [comment]
Just to double check, what does this command say?

```
flatpak run com.github.mtkennerly.ludusavi --version
```

Logging wasn't available before v0.13.0. I know you already specified v0.13.0 in the initial comment, but just making sure the update went through.

Let's also try double checking a few locations:

```
ls -a ~/.var/app/com.github.mtkennerly.ludusavi/config/ludusavi/*
ls -a $XDG_CONFIG_HOME/ludusavi/*
ls -a ~/.config/ludusavi/*

flatpak run --command='ls -a ~/.var/app/com.github.mtkennerly.ludusavi/config/ludusavi/*' com.github.mtkennerly.ludusavi
flatpak run --command='ls -a $XDG_CONFIG_HOME/ludusavi/*' com.github.mtkennerly.ludusavi
flatpak run --command='ls -a ~/.config/ludusavi/*' com.github.mtkennerly.ludusavi
```

**Comment by Bboy486** [comment]
$ flatpak run com.github.mtkennerly.ludusavi --version
ludusavi 0.12.1

 flatpak run com.github.mtkennerly.ludusavi --version
ludusavi 0.12.1


(deck@steamdeck ~)$ ls -a ~/.var/app/com.github.mtkennerly.ludusavi/config/ludusavi/*
/home/deck/.var/app/com.github.mtkennerly.ludusavi/config/ludusavi/config.yaml
/home/deck/.var/app/com.github.mtkennerly.ludusavi/config/ludusavi/manifest.yaml
(deck@steamdeck ~)$ ls -a $XDG_CONFIG_HOME/ludusavi/*
ls: cannot access '/ludusavi/*': No such file or directory


(2)(deck@steamdeck ~)$ ls -a ~/.config/ludusavi/*
ls: cannot access '/home/deck/.config/ludusavi/*': No such file or directory


(2)(deck@steamdeck ~)$ flatpak run --command='ls -a ~/.var/app/com.github.mtkennerly.ludusavi/config/ludusavi/*' com.github.mtkennerly.ludusavi
bwrap: execvp ls -a ~/.var/app/com.github.mtkennerly.ludusavi/config/ludusavi/*: No such file or directory


(1)(deck@steamdeck ~)$ flatpak run --command='ls -a $XDG_CONFIG_HOME/ludusavi/*' com.github.mtkennerly.ludusavi
F: Can't get document portal: GDBus.Error:org.freedesktop.DBus.Error.TimedOut: Failed to activate service 'org.freedesktop.portal.Documents': timed out (service_start_timeout=120000ms)
bwrap: execvp ls -a $XDG_CONFIG_HOME/ludusavi/*: No such file or directory

**Comment by Bboy486** [comment]
(deck@steamdeck ~)$ flatpak run --command='ls -a ~/.config/ludusavi/*' com.github.mtkennerly.ludusavi
bwrap: execvp ls -a ~/.config/ludusavi/*: No such file or directory

**Comment by mtkennerly** [comment]
> ludusavi 0.12.1

Okay, yeah, that's the problem 😅 Please update:

```
flatpak update flathub com.github.mtkennerly.ludusavi
```

And then try again:

```
flatpak run --env=RUST_LOG=ludusavi=trace com.github.mtkennerly.ludusavi backup --preview "Deadpool"
```

**Comment by Bboy486** [comment]
flatpak update flathub com.github.mtkennerly.ludusavi
Looking for updates…
error: Invalid id flathub: Names must contain at least 2 periods

**Comment by mtkennerly** [comment]
Ah, sorry, update has different syntax than install. Try this one:

```
flatpak update com.github.mtkennerly.ludusavi
```

**Comment by Bboy486** [comment]
sudo flatpak update com.github.mtkennerly.ludusavi
[sudo] password for deck:
Looking for updates…


        ID                             Branch Op Remote  Download
 1. [✓] com.github.mtkennerly.ludusavi stable u  flathub 5.8 MB / 5.8 MB

Warning: Not exporting file com.github.mtkennerly.ludusavi.appdata.xml of
unsupported type.

Note that '/var/lib/flatpak/exports/share' is not in the search path
set by the XDG_DATA_HOME and XDG_DATA_DIRS
environment variables, so applications may not
        ID                             Branch Op Remote  Download
 1. [✓] com.github.mtkennerly.ludusavi stable u  flathub 5.8 MB / 5.8 MB

Warning: Not exporting file com.github.mtkennerly.ludusavi.appdata.xml of
unsupported type.
Updates complete.

**Comment by mtkennerly** [comment]
Great! Looks like it updated successfully. I think we're almost there. Now let's try these again:

```
flatpak run com.github.mtkennerly.ludusavi --version
```

```
flatpak run --env=RUST_LOG=ludusavi=trace com.github.mtkennerly.ludusavi backup --preview "Deadpool"
```

Afterwards, please check ~/.var/app/com.github.mtkennerly.ludusavi/config/ludusavi one more time for the log file.

**Comment by Bboy486** [comment]
I tried some other games in that folder 



(deck@steamdeck ~)$ flatpak run --env=RUST_LOG=ludusavi=trace com.github.mtkennerly.ludusavi backup --preview "Hades"
Overall:
  Games: 0
  Size: 0 B
  Location: /run/media/mmcblk0p1/ludusavi
(deck@steamdeck ~)$ flatpak run --env=RUST_LOG=ludusavi=trace com.github.mtkennerly.ludusavi backup --preview "Disco Elysium"
Overall:
  Games: 0
  Size: 0 B
  Location: /run/media/mmcblk0p1/ludusavi

**Comment by Bboy486** [comment]
Here are the logs
[ludusavi_r2022-10-03_00-22-02.log](https://github.com/mtkennerly/ludusavi/files/9693901/ludusavi_r2022-10-03_00-22-02.log)
[ludusavi_r2022-10-03_00-18-29.log](https://github.com/mtkennerly/ludusavi/files/9693902/ludusavi_r2022-10-03_00-18-29.log)

**Comment by mtkennerly** [comment]
Thank you! Here's what I see based on the logs:

### Deadpool
I do see that it found one file for Deadpool:

```
DEBUG [ludusavi::prelude] [Deadpool] found: /home/deck/Games/Deadpool/TransGame/Config/PC/Cooked/Coalesced.ini
```

Do you see that file in the GUI as well now? I believe this was fixed by v0.13.0's removal of the "exclude saves that have only been confirmed on another OS" setting, which I see was enabled in the config file you provided before.

### Hades
PCGamingWiki only has Windows, Mac, and Linux-with-Steam save locations. Since you're using Linux-without-Steam for this game, there are no known save locations currently. In this situation, someone would need to update PCGamingWiki to add the Linux-without-Steam save locations.

### Disco Elysium
The logs for this one are probably in a third file, `ludusavi_rCURRENT.log`. However, looking at PCGamingWiki, it appears to be in the same situation as Hades, so I think that's why it didn't find anything.

**Comment by Bboy486** [comment]
Yup Deadpool shows now when I run preview ony the app. Thank you. 

For the other games can't Ludusavi look into the compdata folders that proton creates? I actually thought that was how you were running the backup since proton is creating the files within the compdata.

**Comment by mtkennerly** [comment]
> Yup Deadpool shows now when I run preview ony the app. Thank you.

Great! It is only the settings, though, not the save data. The only known save location is in the Windows documents folder, which it can't currently find because:

> For the other games can't Ludusavi look into the compdata folders that proton creates?

It can do that for Steam roots, but not for other roots. However, if there's a standard location where Proton puts the compatdata folders without Steam involved, then I can definitely add that. Could you tell me a little more about how you have Hades and Disco Elysium set up with Proton?

* Where are their `drive_c` folders?
* Did you have to manually specify where to put the compatdata, or does Proton decide the location automatically?
* Inside of `drive_c/users`, what usernames are there? Is there a folder for `steamuser`?

**Comment by Bboy486** [comment]
/home/deck/.local/share/Steam/steamapps/compatdata/4272918735/pfx/drive_c/users/steamuser/Documents/Saved Games/Hades/Profile1.sav

**Comment by mtkennerly** [comment]
Ah, that's very interesting. The normal app ID for Hades is 1145360, so I had to do some research about where that 4272918735 ID came from. Apparently, Steam generates app IDs for non-Steam games based on the executable and its location. Ludusavi would need to be able to map Hades <-> 4272918735 in order to find the saves, either by predicting the ID or by reading some Steam config file. I think this may be feasible, but I'll need to look into it more.

**Comment by Bboy486** [comment]
One idea is to have us Steam Deck users provide the IDs for the non steam
games. You could then pull from this list (Google sheets) and hardcode it
into Ludusavi. This is assuming the ID doesn't change numbers.

**Comment by mtkennerly** [comment]
Unfortunately, the ID depends on the exact folder where it's installed. For example, if you have `/home/deck/Games/Hades/hades.exe` and someone else has `/home/deck/drm-free/Hades/hades.exe`, then they'll get different IDs.

Steam has a config file with the non-Steam games that you've added, so it should be possible to pull the info from there, though.

**Comment by Bboy486** [comment]
Awesome. Please let me know what you find out

**Comment by mtkennerly** [comment]
The functionality I would need for this has been added to [steamlocate](https://github.com/WilliamVenner/steamlocate-rs/). Once they publish a new release, I'll be able to add this to Ludusavi \:D

**Comment by Bboy486** [comment]
Thank you. I'll trust one in that since I don't know how that helps.

**Comment by mtkennerly** [comment]
This feature will be included in the next release of Ludusavi (0.15.0). I can't promise an exact date, but it should be in the near future.
